src/numeric/processes.jl

```diff
@@ -8,36 +10,62 @@ abstract DiscreteProcess <: AbstractProcess
 abstract ContinuousProcess <: AbstractProcess
 abstract IIDExogenous <: AbstractProcess
 
-abstract AbstractDiscretizedProcess <: DiscreteProcess
+abstract AbstractDiscretizedProcess # <: DiscreteProcess
 
+###
+### Discretized process
+###
 
-# discretized process
-
+# date-t grid has a known structure
 type DiscretizedProcess <: AbstractDiscretizedProcess
-    nodes::Matrix{Float64}
-    integration_nodes::Array{Matrix{Float64}}
-    integration_weights::Array{Vector{Float64}}
+    grid::Grid
+    integration_nodes::Array{Matrix{Float64},1}
+    integration_weights::Array{Vector{Float64},1}
 end
 
-n_nodes(dp::DiscretizedProcess) = size(dp.nodes, 1)
+n_nodes(dp::DiscretizedProcess) = n_nodes(dp.grid)
+node(dp::DiscretizedProcess, i) = node(dp.grid,i)
 n_inodes(dp::DiscretizedProcess, i::Int) = size(dp.integration_nodes[i], 1)
-inodes(dp::DiscretizedProcess, i::Int, j::Int) = dp.integration_nodes[i][j, :]
-iweights(dp::DiscretizedProcess, i::Int, j::Int) = dp.integration_weights[i][j]
-node(dp::DiscretizedProcess, i) = dp.nodes[i, :]
+inode(dp::DiscretizedProcess, i::Int, j::Int) = dp.integration_nodes[i][j, :]
```

Comment:
So this is the big elephant in the room.
In general terms, it is quite clear that the current choices of data layout are far from optimal. As you correctly note, the strategy is to isolate algorithmic code as much as possible from this decision, hence the functions `inode`, `iweight`, ... The most problematic point right now, is the storage of the computed values or the residuals. Currently it is `Array{Array{Float64,2},1}` which is a half-baked solution. There is also no abstraction on top of it to hide this choice.
As for the data layout of DiscretizedProcess I would advocate instead `dp.integration_nodes[i][j]` to get a vector. Indeed, we don't make anywhere the assumption that the number of integration nodes has to be the same for all states today, which allows for the design of fancy markov process.